aws/session/internal/ini/literal_tokens.go

```diff
@@ -0,0 +1,277 @@
+package ini
+
+import (
+	"fmt"
+	"strconv"
+	"strings"
+	"unicode/utf8"
+
+	"github.com/aws/aws-sdk-go/aws/awserr"
+)
+
+var literalValues = []string{
+	"true",
+	"false",
+}
+
+func isBoolValue(b []byte) bool {
+	for _, lv := range literalValues {
+		if len(b) < len(lv) {
+			continue
+		}
+
+		v := string(b[:len(lv)])
+		if v == lv {
+			return true
+		}
+	}
+	return false
+}
+
+// isNumberValue will return whether not the leading characters in
+// a byte slice is a number. A number is delimited by whitespace or
+// the newline token.
+//
+// A number is defined to be in a binary, octal, decimal (int | float), hex format,
+// or in scientific notation.
+func isNumberValue(b []byte) bool {
+	negativeIndex := 0
+	helper := numberHelper{}
+
+	for i := 0; i < len(b); i++ {
+		negativeIndex++
+
+		switch b[i] {
+		case '-':
+			if helper.IsNegative() || negativeIndex != 1 {
+				return false
+			}
+			helper.Determine(b[i])
+			continue
+		case 'e', 'E':
+			if err := helper.Determine(b[i]); err != nil {
+				return false
+			}
+			negativeIndex = 0
+			continue
+		case 'b':
+			if helper.hex {
+				break
+			}
+			fallthrough
+		case 'o', 'x':
+			if i == 0 {
+				return false
+			}
+
+			fallthrough
+		case '.':
+			if err := helper.Determine(b[i]); err != nil {
+				return false
+			}
+			continue
+		}
+
+		if isWhitespace(b[i]) ||
+			isNewline(b[i:]) {
+			return true
+		}
+
+		if !helper.CorrectByte(b[i]) {
+			return false
+		}
+	}
+
+	return true
+}
+
+func isValid(b []byte) (bool, int, error) {
+	r, size := utf8.DecodeRune(b)
+	if !utf8.ValidRune(r) {
+		return false, 0, awserr.New(ErrCodeParseError, "invalid character", nil)
+	}
+	return isValidRune(r), size, nil
+}
+
+func isValidRune(r rune) bool {
+	return r != '=' && r != '[' && r != ']' && r != ' ' && r != '\n'
+}
+
+// UnionValueType is an enum that will signify what type
+// the UnionValue is
+type UnionValueType int
+
+func (v UnionValueType) String() string {
+	switch v {
+	case NoneType:
+		return "NONE"
+	case DecimalType:
+		return "FLOAT"
+	case IntegerType:
+		return "INT"
+	case StringType:
+		return "STRING"
+	case BoolType:
+		return "BOOL"
+	}
+
+	return ""
+}
+
+// UnionValueType enums
+const (
+	NoneType = UnionValueType(iota)
+	DecimalType
+	IntegerType
+	StringType
+	QuotedStringType
+	BoolType
+)
+
+type literalToken struct {
+	Value UnionValue
+	raw   string
+}
+
+// UnionValue is a union container
+type UnionValue struct {
```

Comment:
does this need to be exported? and move the `literalToken`'s type accessor methods to this type.